### PR TITLE
PD: international opt-in form

### PR DESCRIPTION
**Pull request by breville**
This implements a form for teachers to record their attendance at a non-U.S. workshop, at https://studio.code.org/pd/international_workshop.

international_optin_wip_3

### Existing logged-out page (now loc'ed)

international_optin_wip_3a

### Existing not-teacher page (now loc'ed)

international_optin_wip_3b

### New  teacher-with-no-email page (loc'ed)
international_optin_wip_3c

### Thanks (loc'ed)

international_optin_wip_3d

### CODE REVIEW

File: dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    resources: ['Kodable'],
+    robotics: ['LEGO Education'],
+    workshopOrganizer: 'Workshop Organizer',
+    workshopFacilitator: 'Workshop Facilitator',
+    workshopCourse: 'Workshop Course',
+    optIn: 'Yes'
+  }
+
+  test 'create creates a new international optin' do
+    assert_creates Pd::InternationalOptin do
```
Comment (breville) [inline dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb]: Not sure why this fails...

File: dashboard/app/models/pd/international_optin.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+#
+#  index_pd_international_optins_on_user_id  (user_id)
+#
+
+require 'international_optin_people'
+
+class Pd::InternationalOptin < ApplicationRecord
+  include Pd::Form
+
+  belongs_to :user
```
Comment (breville) [inline dashboard/app/models/pd/international_optin.rb]: Not seeing an associated `user_id` in the database after submitting this form, even when signed in.  Is there something else needed to make the association work?

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                date={date}
+                minDate={moment()}
+                onChange={this.handleDateChange}
+                readOnly={false}
+              />
+            </Col>
+          </Row>
+        </FormGroup>
+        {
+          this.buildSelectFieldGroupFromOptions({
```
Comment (breville) [inline apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: Even though these dropdowns are shown with a default selection, pressing Submit causes them to be highlighted as required, indicating that there's no known selection at this point.  Is there a preferred solution?

File: apps/Gruntfile.js [hunk 4]
```diff
@@ -513,6 +513,8 @@ describe('entry tests', () => {
     'pd/professional_learning_landing/index': './src/sites/studio/pages/pd/professional_learning_landing/index.js',
     'pd/regional_partner_contact/new': './src/sites/studio/pages/pd/regional_partner_contact/new.js',
 
+    'pd/international_optin/new': './src/sites/studio/pages/pd/international_optin/new.js',
```
Comment (aoby) [inline apps/Gruntfile.js]: nit: s/optin/opt_in

And elsewhere. optin isn't a word, nor a common abbreviation AFAIK, so I'd go with `opt_in` and `OptIn` over `optin` and `Optin`

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+  /**
+   * @override
+   */
+  serializeFormData() {
+    const formData = super.serializeFormData();
+    //Object.assign(formData['form_data'], this.getDistrictData());
+
+    return formData;
+  }
```
Comment (aoby) [inline apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: What is the point of this override if it just calls super and returns the result?

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx (hunk 3 above)
Comment (aoby) [reply apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: There is no default value in React component props, and it will render as [empty string](https://github.com/code-dot-org/code-dot-org/blob/b09fe594b804cb674c24237df812279b15415c66/apps/src/code-studio/pd/form_components/FormComponent.jsx#L143) initially. I think at that point, the browser shows the first element selected but it doesn't trigger a React event so doesn't set the value. Either initialize `formData` in the controller to have these initial values, or add a [placeholder](https://github.com/code-dot-org/code-dot-org/blob/b09fe594b804cb674c24237df812279b15415c66/apps/src/code-studio/pd/form_components/FormComponent.jsx#L97-L98) here

File: dashboard/app/models/pd/international_optin.rb (hunk 2 above)
Comment (aoby) [reply dashboard/app/models/pd/international_optin.rb]: The association is fine, but its value is not being set. It should be set on initialization, [here](https://github.com/code-dot-org/code-dot-org/pull/23235/files#diff-14445ae63d23d921b409ef229f562148R3).

Comment (aoby) [reply dashboard/app/models/pd/international_optin.rb]: Is `user_id` required? If so we should also probably update the db column to not null, add a required validation, and controller auth.

File: dashboard/config/locales/en.yml [hunk 6]
```diff
@@ -1021,6 +1021,81 @@ en:
       things_you_liked_s: 'What were the two things you liked most about the activities you did in this workshop and why?'
       things_you_would_change_s: 'What are the two things you would change about the activities you did in this workshop?'
       anything_else_s: 'Is there anything else you’d like to tell us about your experience at this workshop?'
+    form_labels:
```
Comment (aoby) [inline dashboard/config/locales/en.yml:1024]: I think @tanyaparker may have suggested putting this in a new file for forms i18n?

File: dashboard/db/migrate/20180620110434_create_pd_international_optins.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class CreatePdInternationalOptins < ActiveRecord::Migration[5.0]
+  def change
+    create_table :pd_international_optins do |t|
+      t.references :user, index: true
+      t.text :form_data
```
Comment (aoby) [inline dashboard/db/migrate/20180620110434_create_pd_international_optins.rb]: Is user required?
Presumably form_data is required, and should be not null.

Also please move the migration out into its own PR (per convention, in case we have to roll back).

File: dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb (hunk 1 above)
Comment (aoby) [reply dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb]: Perhaps the API call is failing for some reason? I'd suggest moving the `assert_response` inside the `assert_creates` so we first assert the successful response and then assert that it created a new model instance.

Also does `assert_response :created` (201) work? I think we return 200 `:success`. In any case, I suspect we're failing to create and returning a 400 or other error code. See https://www.rubydoc.info/gems/rack/Rack/Utils for the list of codes.

File: dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    new_form = SAMPLE_FORM_DATA.dup
+    new_form.delete :lastName
+
+    assert_does_not_create Pd::InternationalOptin do
+      put :create, params: {
+        form_data: new_form
+      }
+    end
+
+    assert_response :bad_request
```
Comment (aoby) [inline dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb]: Same as above, move this inside the `assert_creates`. 

I think it's a clearer test (and easier to decipher results) to first check the return code, then check the side effects.

File: dashboard/test/models/pd/international_optin_test.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    workshopFacilitator: 'Workshop Facilitator',
+    workshopCourse: 'Workshop Course',
+    optIn: 'Yes'
+  }
+
+  test 'Test international optin validation' do
+    optin = build :pd_international_optin, form_data: {}.to_json
+    refute optin.valid?
+
+    assert build(:pd_international_optin, form_data: FORM_DATA.to_json).valid?
```
Comment (aoby) [inline dashboard/test/models/pd/international_optin_test.rb]: nit: why 2 lines for refute and one line for assert?

The `optin` variable from line 25 doesn't appear to be used anywhere else (after line 26, which can be combined).

File: lib/international_optin_people.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
+INTERNATIONAL_OPTIN_FACILITATORS = [
```
Comment (aoby) [inline lib/international_optin_people.rb]: This should be contained in a named module, which can then be included where needed. 

Then, since `dashboard/lib` is [autoloaded](https://github.com/code-dot-org/code-dot-org/blob/bd5abfe88ff4e4f3eb8d2949c100b34c8d95f12b/dashboard/config/application.rb#L116) we won't need to explicitly [require](https://github.com/code-dot-org/code-dot-org/pull/23235/files#diff-c89624673ce6bf54fdb9c2f8cbb32055R16) this file, and these constants and methods will be contained in the module and applied to the imported class. Better containment, better style :)

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+import moment from 'moment';
+import {DATE_FORMAT} from '../workshop_dashboard/workshopConstants';
+import {
+  Row,
+  Col,
+  ControlLabel,
+  FormGroup
+} from 'react-bootstrap';
+
+export default class InternationalOptin extends FormController {
```
Comment (mehalshah) [inline apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: Suggest renaming to `InternationalOptIn`

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+      accountEmail: this.props.accountEmail,
+      labels: this.props.labels
+    };
+  }
+}
+
+
+class InternationalOptinComponent extends FormComponent {
+  static propTypes = {
+    accountEmail: PropTypes.string.isRequired
```
Comment (mehalshah) [inline apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: Are we going to pre-fill the name as well?

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+            type: 'text',
+            required: true
+          })
+        }
+        {
+          this.buildFieldGroup({
+            name: 'email',
+            label: labels.email,
+            type: 'text',
+            value: this.props.accountEmail,
```
Comment (mehalshah) [inline apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: Thinking about this a bit - there are some users that would fill out this form that do not have an email (they mistakenly created a student account way back when). Maybe make this readOnly only if we have an email?

File: dashboard/app/controllers/api/v1/pd/international_optins_controller.rb [hunk 14]
```diff
@@ -0,0 +1,17 @@
+class Api::V1::Pd::InternationalOptinsController < Api::V1::Pd::FormsController
+  def new_form
+    @contact_form = ::Pd::InternationalOptin.new
```
Comment (mehalshah) [inline dashboard/app/controllers/api/v1/pd/international_optins_controller.rb]: Do we let anyone use this form, or only teachers?

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx (hunk 13 above)
Comment (breville) [reply apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: Good point.  We will actually route them to a separate page directing them to edit their profile and add an email, similar to what we do for non-teachers and signed-out users.  So anybody who actually gets to this form will have an email.

File: dashboard/app/controllers/api/v1/pd/international_optins_controller.rb (hunk 14 above)
Comment (breville) [reply dashboard/app/controllers/api/v1/pd/international_optins_controller.rb]: Only teachers with email addresses.

File: dashboard/config/locales/en.yml (hunk 6 above)
Comment (breville) [reply dashboard/config/locales/en.yml:1024]: Yes.  It does add a bunch of translated files, so next step was to ask @Hamms and @nkiruka what they think would be best :)

File: dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb (hunk 1 above)
Comment (breville) [reply dashboard/test/controllers/api/v1/pd/international_optins_controller_test.rb]: Thanks for the suggestion, it helped narrow down the problems (which were a failure to name the class properly and a failure to underscore the parameters).

File: dashboard/test/models/pd/international_optin_test.rb (hunk 9 above)
Comment (breville) [reply dashboard/test/models/pd/international_optin_test.rb]: Good catch; will simplify.

No particular reason, was just basing it off of [these tests](https://github.com/code-dot-org/code-dot-org/blob/394262a0158765dbb4ab4276ea3c1a23455d34e9/dashboard/test/models/pd/regional_partner_contact_test.rb#L25-L28).

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx (hunk 12 above)
Comment (breville) [reply apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: Good idea.  Is teacher name always available as separate first & last name fields on `User`?  Is that what `ops_first_name` and `ops_last_name` are?

Comment (aoby) [reply apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: No, `User.name` is whatever the user provides for Display Name in account settings. 

`ops_first_name` and `ops_last_name` were manually supplied by our internal "ops" team via the old / deprecated ops workshop dashboard. These are not reliable and not up to date.

**Comment by tanyaparker** [comment]
cc @bencodeorg

File: apps/src/code-studio/pd/international_optin/InternationalOptin.jsx (hunk 12 above)
Comment (breville) [reply apps/src/code-studio/pd/international_optin/InternationalOptin.jsx]: Good to know.  We won't pre-fill the name then, since we're not sure how to parse the Display Name afaik.

File: dashboard/config/locales/en.yml (hunk 6 above)
Comment (breville) [reply dashboard/config/locales/en.yml:1024]: @Hamms What do you think?  Is it worth making a new `.yml` file for this form?

File: apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+
+  render() {
+    const labels = this.props.labels;
+    const date = (this.props.data && this.props.data.date) ?
+      moment(this.props.data.date, DATE_FORMAT) : moment();
+    // todo: add validation
+
+    let textFieldMapSubjects = {};
+    const lastSubjectsKey = this.props.options.subjects.slice(-1)[0];
+    textFieldMapSubjects[lastSubjectsKey] = "other";
```
Comment (breville) [inline apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: I didn't want to change the underlying form system, but it seems this would be more robust if we identified the fields based on an identifier rather than by a user-facing string, especially if we're loc'ing like here.

File: dashboard/config/locales/en.yml (hunk 6 above)
Comment (Hamms) [reply dashboard/config/locales/en.yml:1024]: Hmmm, that's a good question. I certainly wouldn't want to create a new i18n file just for this form, but I could definitely see making a new i18n file for form content in general. Do you have a sense of how much content would end up in that file and/or how many new forms of this type we intend to make in the future?

File: apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx (hunk 15 above)
Comment (aoby) [reply apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: Can you elaborate? I'm not sure I understand. Do you mean here the key (option text, which does appear to be localized since it's pulled from the dynamic last value of `props.options.subjects`), or the value (which will become the key suffix for the additional text field)? If the latter, can you explain why localizing the key would be helpful? If the former, what is missing? If you want it to be a constant rather than the last in options, then by all means create a new object mapping field name to the "Other" text fields and pass it through from the server in script data.

Comment (breville) [reply apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: Oh, it's the key.  It just seemed a bit delicate to have a key in `textFieldMap` being a human-readable string (e.g. `"Other:"`, especially since it will vary with language.  As far as I can tell, this is implemented in [FormComponent](https://github.com/code-dot-org/code-dot-org/blob/f0d1496414f321d47c18ccf962bac3b483836c3d/apps/src/code-studio/pd/form_components/FormComponent.jsx#L271-L277) and I didn't want to change that code.  Hopefully not a big deal, unless there's something unexpected in any translations...

Comment (aoby) [reply apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: The human-readable string is the option (out of `options` passed through from the model) that gets an additional text box. It's just used for matching purposes at render time, not stored. The right-side, the value in this mapping, becomes the suffix on the new key in the resulting `form_data` for the new field. Like other field names, this key is not (and should not be) localized.

Hopefully this explanation helps (from just above your link in `FormComponent`): 
https://github.com/code-dot-org/code-dot-org/blob/f0d1496414f321d47c18ccf962bac3b483836c3d/apps/src/code-studio/pd/form_components/FormComponent.jsx#L234-L236

If not, feel free to modify / rewrite to make it more clear.

In our case here with localized `options` the key in this mapping, the option to match, must also be localized. 

Backing up, another design choice originally would have been to have richer objects for `options` that specified aspects of the UI including whether that particular option had an additional text box. However we added this feature after options was already just a simple array, so we chose this approach instead - to specify additional text fields as an add-on, with a mapping from the options that already existed.

Anyway, if both `options` and the source for the `textFieldMap` both come from the model and are localized, then what is delicate? It's only really different from the above approach in that it's distributed over 2 lists.

Comment (breville) [reply apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: We discussed offline.  My main concern was having JS hash key names be user-facing strings that are localised, meaning they can contain special & international characters.  However, we believe that all of those are valid, so this should be fine.

File: dashboard/config/locales/en.yml (hunk 6 above)
Comment (breville) [reply dashboard/config/locales/en.yml:1024]: I'm not aware of any more loc'ed forms that are currently planned to be implemented in this system.  We could certainly extract these strings if/when that time comes, right?

File: apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+    let textFieldMapResources = {};
+    const lastResourcesKey = this.props.options.resources.slice(-1)[0];
+    textFieldMapResources[lastResourcesKey] = "other";
+
+    let textFieldMapRobotics = {};
+    const lastRoboticsKey = this.props.options.robotics.slice(-1)[0];
+    textFieldMapRobotics[lastRoboticsKey] = "other";
```
Comment (aoby) [inline apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: nit: this can be const since its not reassigned (let vs const is about the reference. You can still modify the object contents in an object referred to by a const variable)`

Also the declaration and assignment can be combined if you want:
```jsx
const textFieldMapRobotics = {[lastRoboticsKey]: "other"};
```
See Computed Property Names here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer

File: lib/international_opt_in_people.rb [hunk 17]
```diff
@@ -0,0 +1,13 @@
+module InternationalOptInPeople
```
Comment (aoby) [inline lib/international_opt_in_people.rb]: Is this only used in `Pd::InternationalOptIn`? Why does it need to be its own module? And why in the root `lib` directory? Do we expect to share with Pegasus? 

As [suggested](https://github.com/code-dot-org/code-dot-org/pull/23235#discussion_r198296006) earlier, if we put it instead in `dashboard/lib` we won't need [this require](https://github.com/code-dot-org/code-dot-org/pull/23235/files#diff-b940bb0be51d6330a2b3e0a9f53354feR16).

**Review by aoby — APPROVED** [review]

**Comment by bencodeorg** [comment]
Sorry @breville I forgot about this PR and just checked it out, two questions:

1) can only teacher accounts fill out this form?
2) do you have to be logged in to fill it out?

Asking b/c I believe this form is the way we'll establish teacher "cohorts" for international partners, so any work we can do up front to avoid problems later are appreciated.

**Comment by aoby** [comment]
@bencodeorg Yes and yes, because of the [ability](https://github.com/code-dot-org/code-dot-org/pull/23235/files#diff-38d6349150d685325a41324b3032884eR96) granted only to teachers, and [authorized](https://github.com/code-dot-org/code-dot-org/pull/23235/files#diff-4b537f52972c1aadd69ef0254684fcc0R2) in the controller.

@breville since this is a requirement, it's probably worth adding tests for the `InternationalOptInsController` to make sure that signed in teachers can see the page, but students and sign outs can't. It's probably also worth adding an authorize to the API controller (& tests) for the same.

**Comment by bencodeorg** [comment]
Helpful, thanks @aoby!

**Comment by breville** [comment]
@aoby I added tests that the correct page is rendered in https://github.com/code-dot-org/code-dot-org/pull/23603, though they are for `InternationalOptInController` rather than `InternationalOptInsController`.  Is that right?

`Api::V1::Pd::InternationalOptInsController` already has an `authorize_resource`.  Should `Pd::InternationalOptInController` get one too?

Sorry if I've got these back-to-front.

**Comment by aoby** [comment]
Heh, the names are confusing. How about this: 
 1. The front end controller (`Pd::InternationalOptInController`). Now that I see the tests in https://github.com/code-dot-org/code-dot-org/pull/23603 it's clear this should not have an authorize because it renders various templates for users who are not authorized to see the form. They should not be forbidden. 👍 
 2. The API controller (`Api::V1::Pd::InternationalOptInController`), which does use `authorize_resource` still should have a unit test confirming that authorization IMO - that only (logged in) teachers can create one, while students cannot.
